packages/pigeon/lib/cpp_generator.dart

```diff
@@ -934,22 +1055,33 @@ class CppSourceGenerator extends StructuredGenerator<CppOptions> {
           'flutter::ByteStreamWriter* stream',
         ],
         isConst: true, body: () {
-      if (customTypes.isNotEmpty) {
+      if (enumeratedTypes.isNotEmpty) {
         indent.write(
             'if (const CustomEncodableValue* custom_value = std::get_if<CustomEncodableValue>(&value)) ');
         indent.addScoped('{', '}', () {
-          for (final EnumeratedType customType in customTypes) {
+          for (final EnumeratedType customType in enumeratedTypes) {
+            final String encodeString = customType.type ==
+                    CustomTypes.customClass
+                ? 'std::any_cast<${customType.name}>(*custom_value).ToEncodableList()'
+                : 'static_cast<int>(std::any_cast<${customType.name}>(*custom_value))';
+            final String valueString = customType.enumeration <
+                    maximumCodecFieldKey
+                ? encodeString
+                : 'std::any_cast<$_overflowClassName>(wrap).ToEncodableList()';
+            final int enumeration =
+                customType.enumeration < maximumCodecFieldKey
+                    ? customType.enumeration
+                    : maximumCodecFieldKey;
             indent.write(
                 'if (custom_value->type() == typeid(${customType.name})) ');
             indent.addScoped('{', '}', () {
-              indent.writeln('stream->WriteByte(${customType.enumeration});');
-              if (customType.type == CustomTypes.customClass) {
-                indent.writeln(
-                    'WriteValue(EncodableValue(std::any_cast<${customType.name}>(*custom_value).ToEncodableList()), stream);');
-              } else if (customType.type == CustomTypes.customEnum) {
+              indent.writeln('stream->WriteByte($enumeration);');
+              if (enumeration == maximumCodecFieldKey) {
                 indent.writeln(
-                    'WriteValue(EncodableValue(static_cast<int>(std::any_cast<${customType.name}>(*custom_value))), stream);');
+                    'const auto& wrap = $_overflowClassName(${customType.enumeration - maximumCodecFieldKey}, $encodeString);');
```

Comment:
This one would be better as `auto` rather than `auto&`. The latter relies on subtle lifetime extension rules, and doesn't add anything.